compiler/packages/babel-plugin-react-compiler/src/HIR/CollectHoistablePropertyLoads.ts

```diff
@@ -0,0 +1,436 @@
+    lvalue: Place,
+    object: Place,
+    propertyName: string,
+    shouldDeclare: boolean
+  ): PropertyLoadNode {
+    // temporaries contains object if this is a property load chain from a named variable
+    // Otherwise, there is a non-trivial expression
+    const resolvedObject =
+      this.temporaries.get(object.identifier) ?? object.identifier;
+
+    const resolvedDependency = this.properties.get(resolvedObject);
+    let property: ReactiveScopeDependency;
+    if (resolvedDependency == null) {
+      property = {
+        identifier: resolvedObject,
+        path: [propertyName],
+      };
+    } else {
+      property = {
+        identifier: resolvedDependency.identifier,
+        path: [...resolvedDependency.path, propertyName],
+      };
+    }
+
+    if (shouldDeclare) {
+      this.properties.set(lvalue.identifier, property);
+    }
+    return this.tree.add(property);
+  }
+}
+
+function collectNodes(
+  fn: HIRFunction,
+  functionExprRvals: Set<IdentifierId>,
+  usedOutsideDeclaringScope: Set<IdentifierId>,
+  c: TemporariesSideMap
+): Map<BlockId, BlockInfo> {
+  const nodes = new Map<BlockId, BlockInfo>();
+  const scopeStartBlocks = new Map<BlockId, ScopeId>();
+  for (const [blockId, block] of fn.body.blocks) {
+    const assumedNonNullObjects = new Set<PropertyLoadNode>();
+    for (const instr of block.instructions) {
+      const { value, lvalue } = instr;
+      const usedOutside = usedOutsideDeclaringScope.has(lvalue.identifier.id);
+      if (value.kind === "PropertyLoad") {
+        const propertyNode = c.declareProperty(
+          lvalue,
+          value.object,
+          value.property,
+          !usedOutside
+        );
+        if (
+          !functionExprRvals.has(lvalue.identifier.id) &&
+          !isMutable(instr, value.object)
+        ) {
+          let curr = propertyNode.parent;
+          while (curr != null) {
+            assumedNonNullObjects.add(curr);
+            curr = curr.parent;
+          }
+        }
+      } else if (value.kind === "LoadLocal") {
+        if (
+          lvalue.identifier.name == null &&
+          value.place.identifier.name !== null &&
+          !usedOutside
+        ) {
+          c.declareTemporary(lvalue.identifier, value.place.identifier);
+        }
+      }
+      /**
+       * Note that we do not record StoreLocals as this runs after ExitSSA.
+       * As a result, an expression like `(a ?? b).c` is represented as two
+       * StoreLocals to the same identifier id.
+       */
+    }
+
+    if (
+      block.terminal.kind === "scope" ||
+      block.terminal.kind === "pruned-scope"
+    ) {
+      scopeStartBlocks.set(block.terminal.block, block.terminal.scope.id);
+    }
+
+    nodes.set(blockId, {
+      blockId,
+      scope: scopeStartBlocks.get(blockId) ?? null,
+      preds: block.preds,
+      assumedNonNullObjects,
+    });
+  }
+  return nodes;
+}
+
+function deriveNonNull(fn: HIRFunction, nodes: Map<BlockId, BlockInfo>): void {
```

Comment:
The bookkeeping/traversal logic is making it really hard to follow the core logic of this pass. I would have expected that a non-null object analysis could use the dominator tree rather than the walking forward/backwards that's happening here. Did you try using dominators? If yeah, what was the challenge?

Reply:
From my understanding (please correct if I'm misunderstanding something, my algo knowledge is much weaker than yours), dominator tree would work but misses some cases the non-hir implementation currently covers.

Take this example. Here, we know that it's safe to access `a.b` on line 5 even though no dominator of its basic block accesses a.b.
```js
0: if(...) {
1:   a.b;
2: } else {
3:   a.b;
4: }
5: // It's safe to take `a.b` as a dependency of scope@0 (i.e. hoist the a.b access)
6: scope @0 {
7:   if (cond) {
8:     // access a.b;
9:   }
10:}
```

I believe a more precise formation would be that `a.b` is safe to read in $bb_i$ if any of the following hold:
- $bb_i$ itself accesses a.b
- a.b is safe to access in every pred of `a.b`
- a.b is safe to access in every successor of `a.b`

Basic block cycles can be handled by filtering / dropping the offending block (e.g. reducing `x = x && y && ...` to `x = y && ...`)

Reply:
Gotcha. That makes sense. I suspect this can still be simplified - in particular, i had to remap the variable names in my head to help keep track of things. See comments for the most unclear cases.

One thing that strikes me about this algorithm is that it isn't accurate in the sense that it doesn't account for the actual live range of values. During the initial collectNodes stage we account for mutable ranges when deciding if a property access can locally be considered to be non-nullable. But then in the deriveNonNull phase we don't have access to the mutable ranges anymore. AFAICT, if you have:

```
bb0:
  let x = null
  if cond then bb1 else bb2
bb1:
  x = {}
  goto bb3
bb2:
  ...
  goto bb3:
bb3:
  x.foo.bar
```

this algorithm will infer `x.foo.bar` as non-nullable for bb0, via the intersection of bb1 & bb2 which in turn comes from bb3. So the results are actually inaccurate, but that's okay because of how we filter out dependencies in PropagateScopeDeps. 

This isn't necessarily a blocker, but it isn't great to have a pass that returns results that are technically incorrect — it's very easy to accidentally use them in a different way later, w/o realizing that important context. Thoughts on this, and whether we could make the results guaranteed accurate?